File: client/src/main/java/com/jraska/github/client/push/PushIntentObserver.java
package com.jraska.github.client.push;

import android.app.Activity;
import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.LifecycleObserver;
import android.arch.lifecycle.LifecycleOwner;
import android.arch.lifecycle.OnLifecycleEvent;
import android.content.Intent;

import com.google.firebase.messaging.RemoteMessage;
import com.google.firebase.messaging.RemoteMessageBridge;

public final class PushIntentObserver implements LifecycleObserver {
  private final PushHandler pushHandler;

  public PushIntentObserver(PushHandler pushHandler) {
    this.pushHandler = pushHandler;
  }

  @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
  public void onCreate(LifecycleOwner owner) {
    if (!(owner instanceof Activity)) {
      return;
    }

    Activity activity = (Activity) owner;
    Intent intent = activity.getIntent();
    if (isPush(intent)) {
      RemoteMessage message = RemoteMessageBridge.INSTANCE.create(intent.getExtras());
      PushAction action = RemoteMessageToActionConverter.INSTANCE.convert(message);

      pushHandler.handlePush(action);
    }
  }

  private boolean isPush(Intent intent) {
    if (!Intent.ACTION_MAIN.equals(intent.getAction()) || intent.getExtras() == null) {
      return false;
    }

    for (String key : intent.getExtras().keySet()) {
      if (key.startsWith("google.sent_time")) {
        return true;
      }

      if (key.startsWith("google.to")) {
        return true;
      }

      if (key.startsWith("gcm.")) {
        return true;
      }
    }

    return false;
  }
}

File: client/src/main/java/com/jraska/github/client/push/PushIntentObserver.kt
package com.jraska.github.client.push

import android.app.Activity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.OnLifecycleEvent
import android.content.Intent
import com.google.firebase.messaging.RemoteMessage

class PushIntentObserver(private val pushHandler: PushHandler) : LifecycleObserver {

  @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
  fun onCreate(owner: LifecycleOwner) {
    if (owner !is Activity) {
      return
    }

    val activity = owner as Activity
    val intent = activity.intent
    if (isPush(intent)) {
      val message = RemoteMessage(intent.extras!!)
      val action = RemoteMessageToActionConverter.convert(message)

      pushHandler.handlePush(action)
    }
  }

  private fun isPush(intent: Intent): Boolean {
    if (Intent.ACTION_MAIN != intent.action || intent.extras == null) {
      return false
    }

    for (key in intent.extras!!.keySet()) {
      if (key.startsWith("google.sent_time")) {
        return true
      }

      if (key.startsWith("google.to")) {
        return true
      }

      if (key.startsWith("gcm.")) {
        return true
      }
    }

    return false
  }
}
